src/Core/File.cpp

```diff
@@ -0,0 +1,162 @@
+
+namespace Acamarachi::Core
+{
+
+    ErrorOr<File> File::open(const char *filepath, OpenMode mode)
+    {
+        File f = {};
+        f.fd = _open(filepath, (int)mode);
+        if (f.fd == -1)
+        {
+            switch (errno)
+            {
+            case EACCES:
+                return ErrorCode::AccessDenied;
+            case EEXIST:
+                return ErrorCode::PathAlreadyExists;
+            case EISDIR:
+                return ErrorCode::IsDirectory;
+            case ELOOP:
+                return ErrorCode::SymbolicLinkLoop;
+            case EMFILE:
+                return ErrorCode::ProcessFdQuotaExceeded;
+            case ENAMETOOLONG:
+                return ErrorCode::NameTooLong;
+            case ENFILE:
+                return ErrorCode::SystemFdQuotaExceeded;
+            case ENOENT:
+                return ErrorCode::FileNotFound;
+            case ENOSPC:
+                return ErrorCode::NoSpaceLeft;
+            case EOVERFLOW:
+                return ErrorCode::Overflow;
+            case EROFS:
+                return ErrorCode::ReadOnlyFileSystem;
+            case EAGAIN:
+                return ErrorCode::WouldBlock;
+            case ENOMEM:
+                return ErrorCode::OutOfMemory;
+            case ETXTBSY:
+                return ErrorCode::TextBusy;
+            default:
+                return ErrorCode::Unexpected;
+            }
+        }
+        auto maybe_size = f.getFileSize();
+        if (!maybe_size)
+        {
+            return maybe_size.error();
+        }
+        f.size = maybe_size.result();
+        f.mode = mode;
+        return f;
+    }
+
+    ErrorOr<File> File::open(const std::string &filepath, OpenMode mode)
+    {
+        return open(filepath.c_str(), mode);
+    }
+
+    ErrorOr<ssize_t> File::read(char *buffer, size_t n)
+    {
+        assert(mode != File::OpenMode::Write);
+        assert(fd != -1);
+        ssize_t r = cRead(fd, buffer, (unsigned int)n);
+        if (r == -1)
+        {
+            switch (errno)
+            {
+            case EFAULT:
+                return ErrorCode::PointsOutOfMemory;
+            default:
+                return ErrorCode::Unexpected;
+            }
+        }
+        return r;
+    }
+
+    ErrorOr<ssize_t> File::read(Slice<char> &slice)
+    {
+        assert(mode != File::OpenMode::Write);
+        assert(fd != -1);
+        ssize_t r = cRead(fd, slice.ptr, (unsigned int)slice.len);
+        if (r == -1)
+        {
+            switch (errno)
+            {
+            case EFAULT:
+                return ErrorCode::PointsOutOfMemory;
+            default:
+                return ErrorCode::Unexpected;
+            }
+        }
+        return r;
+    }
+
+    // Using mingw64 read replace '\n\r' to '\n', but mingw64 stat counts '\r' in the file size.
+    // This lead to over allocation of memory, 1 byte per newline.
+    ErrorOr<Slice<char>> File::readAll(Allocator::Interface allocator)
```

Comment:
will test under MSVC, migth output specific error also.

Reply:
We might have to use some preprocessor to harmonize the behaviour if there is a difference

Reply:
yeah, i'll sort MSVC stuff!